conbench/entities/history.py

```diff
@@ -570,3 +583,82 @@ def _calculate_z_score(
         z_score = z_score * -1
 
     return z_score
+
+
+class _AutomatedDetector(ABC):
+    @abstractmethod
+    def detect(self, df: pd.DataFrame) -> pd.Series:
+        """Detect outliers and distribution shifts in historical data
+
+        For a particular detection algorithm, should take a dataframe of the same
+        sort as `_add_rolling_stats_columns_to_df`, and return that dataframe with
+        two additional columns:
+
+        - `is_step` (bool): Is this point the start of a new segment?
+        - `is_outlier` (bool): Is this point an outlier that should be ignored?
+        """
+
+
+class _ZScoreAutomatedDetector(_AutomatedDetector):
+    def detect(self, df: pd.DataFrame, z_score_threshold=5.0) -> pd.DataFrame:
+        tmp_df = copy.deepcopy(df)
+
+        # skip computation if no history
+        if df.shape[0] == 0:
+            tmp_df["is_step"] = pd.Series([], dtype=bool)
+            tmp_df["is_outlier"] = pd.Series([], dtype=bool)
+            return tmp_df
+
+        # pandas likes the data to be sorted
+        tmp_df.sort_values(
+            [
+                "case_id",
+                "context_id",
+                "hash",
+                "repository",
+                "timestamp",
+                "result_timestamp",
+            ],
+            inplace=True,
+            ignore_index=True,
+        )
+
+        # split / apply
+        out_group_df_list = []
+        for _, group_df in tmp_df.groupby(
+            ["case_id", "context_id", "hash", "repository"]
+        ):
+            # clean copy will only get result columns
+            out_group_df = copy.deepcopy(group_df)
+
+            group_df["mean_diff"] = group_df["mean"].diff()
+            mean_diff_clipped = group_df.mean_diff[
+                (group_df.mean_diff > group_df.mean_diff.quantile(0.05))
+                & (group_df.mean_diff < group_df.mean_diff.quantile(0.95))
+            ]
+            group_df["z_score"] = (
+                group_df.mean_diff - mean_diff_clipped.mean()
+            ) / mean_diff_clipped.std()
+
+            group_df["is_shift"] = group_df.z_score.abs() > z_score_threshold
+            group_df["reverts"] = (
+                group_df.is_shift
+                & group_df.is_shift.shift(-1)
```

Comment:
This does look one observation into the future because it has to to see if it reverts. We obviously don't know if the last value reverts or not.